daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -174,61 +182,146 @@ private[lf] final class ValueTranslator(
-                val fields = labeledRecordToMap(flds) match {
-                  case None =>
-                    (recordFlds zip flds).map { case ((lbl, typ), (mbLbl, v)) =>
-                      mbLbl.foreach(lbl_ =>
-                        if (lbl_ != lbl)
-                          typeError(
-                            s"Mismatching record field label '$lbl_' (expecting '$lbl') for record $tyCon"
-                          )
-                      )
-                      val replacedTyp = AstUtil.substitute(typ, subst)
-                      lbl -> go(replacedTyp, v, newNesting)
+
+                if (config.enableUpgrade) {
+                  // We handle the (non-)upgrade cases for record separately
+                  // In case of upgrade, we require the field are record.
+
+                  // This code implements the compatibility transformation used for up/down-grading
+                  // And handles the cases:
+                  // - UPGRADE:   numT > numS : creates a None for each missing fields.
+                  // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
+                  //
+                  // When numS == numT, we wont hit the code marked either as UPGRADE or DOWNGRADE,
+                  // although it is still possible that the source and target types are different,
+                  // but since we don't consult the source type (may be unavailable), we wont know.
+
+                  val numS: Int = sourceElements.length
+                  val numT: Int = targetFieldsAndTypes.length
+
+                  // traverse the sourceElements, "get"ing the corresponding target type
+                  // when there is no corresponding type, we must be downgrading, and so we insist the value is None
+                  val values0: List[SValue] =
+                    sourceElements.toSeq.view.zipWithIndex.flatMap { case ((optName, v), i) =>
+                      targetFieldsAndTypes.get(i) match {
+                        case Some((targetField, targetFieldType)) =>
+                          optName match {
+                            case None => ()
+                            case Some(sourceField) =>
+                              // value is not normalized; check field names match
+                              if (sourceField != targetField)
+                                typeError(
+                                  s"Mismatching record field label '$sourceField' (expecting '$targetField') for record $tyCon"
+                                )
+
+                          }
+                          val typ: Type = AstUtil.substitute(targetFieldType, subst)
+                          val sv: SValue = go(typ, v, newNesting)
+                          List(sv)
+                        case None => // DOWNGRADE
+                          // i ranges from 0 to numS-1. So i >= numT implies numS > numT
+                          assert((numS > i) && (i >= numT))
```

Comment:
This assertion is for documentation purposes I guess as it should be guaranteed externally by the `ImmArray` API

Reply:
I assume yes. 
This is a copy/paste of the code by @nickchapman-da  in Speedy.scala